File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        max_length = 0
        seen = set()
        left = 0
        for right in range(len(s)):
            while(s[right] in seen and right < len(s)):
                seen.remove(s[left])
                left += 1
            seen.add(s[right])
            max_length = max(max_length, right-left+1)
        return max_length

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> charSeen = new HashSet<>();
        int left = 0;
        int max_length = 0;
        for(int right=0; right < s.length(); right++) {
            while(!charSeen.add(s.charAt(right))) {
                charSeen.remove(s.charAt(left));
                left++;
            }
            charSeen.add(s.charAt(right));
            max_length = Math.max(max_length, right-left+1);
        }
        return max_length;
    }
}